### PR TITLE
feat/data-upload: add /upload for blank records

### PR DESCRIPTION
**Pull request by rudyardrichter**
### New Features

Adds `/data/upload` endpoint for the data upload flow which generates a "blank" record in indexd and returns a signed URL for uploading data to S3, along with the GUID.

### Breaking Changes

None

### Bug Fixes

None

### Improvements

Moved data blueprint into its own directory and did some refactoring of the existing code & tests for data upload/download. Removed a bit of gunk in the tests by using `@cached_property` to eliminate some methods on `IndexedFile`.

### Dependency updates

Added dependency
- `cached_property>=1.5.1`

### Deployment changes

None

### CODE REVIEW

**Comment by PlanXCyborg** [comment]
This PR contains code that is not formatted correctly according to [`black`](https://github.com/ambv/black). Run `black` on your code before merging.

<details>
<summary>Expand the full diff to see formatting changes</summary>

```diff
--- fence/blueprints/data/__init__.py
+++ blackened
@@ -1,7 +1,5 @@
 from fence.blueprints.data.blueprint import blueprint
 
 
-__all__ = [
-    "blueprint",
-]
+__all__ = ["blueprint"]
 

--- fence/blueprints/data/indexd.py
+++ blackened
@@ -20,11 +20,11 @@
 from fence.errors import (
     InternalError,
     NotFound,
     NotSupported,
     Unauthorized,
-    UnavailableError ,
+    UnavailableError,
 )
 from fence.resources.google.utils import (
     get_or_create_primary_service_account_key,
     create_primary_service_account_key,
     get_or_create_proxy_group_id,
@@ -171,12 +171,13 @@
                     "indexd response missing JSON field {}".format(url + self.file_id)
                 )
                 raise InternalError("internal error from indexd: {}".format(e))
         elif res.status_code == 404:
             flask.current_app.logger.error(
-                "Not Found. indexd could not find {}: {}"
-                .format(url + self.file_id, res.text)
+                "Not Found. indexd could not find {}: {}".format(
+                    url + self.file_id, res.text
+                )
             )
             raise NotFound("No indexed document found with id {}".format(self.file_id))
         else:
             raise UnavailableError(res.text)
 

--- setup.py
+++ blackened
@@ -8,12 +8,11 @@
         "oauth2client<4.0dev,>=2.0.0",
         "addict>=2.1.1, <3.0.0",
         "boto>=2.36.0,<3.0.0",
         "botocore>=1.7,<1.9.0",
         "boto3>=1.5,<1.6",
-        "cached_property>=1.5.1,<2.0.0"
-        "cirrus",
+        "cached_property>=1.5.1,<2.0.0" "cirrus",
         "cryptography>=2.1.2",
         "enum34>=1.1.6",
         "flask-restful>=0.3.6,<1.0.0",
         "Flask>=0.10.1,<1.0.0",
         "Flask-CORS>=3.0.3,<4.0.0",

--- tests/conftest.py
+++ blackened
@@ -406,12 +406,11 @@
             "updated_date": "",
         }
 
     # TODO (rudyardrichter, 2018-11-03): consolidate things needing to do this patch
     indexd_patcher = patch(
-        "fence.blueprints.data.indexd.IndexedFile.index_document",
-        record,
+        "fence.blueprints.data.indexd.IndexedFile.index_document", record
     )
     mocker.add_mock(indexd_patcher)
 
     output = {
         "mocker": mocker,
@@ -483,12 +482,11 @@
             "created_date": "",
             "updated_date": "",
         }
 
     indexd_patcher = patch(
-        "fence.blueprints.data.indexd.IndexedFile.index_document",
-        record,
+        "fence.blueprints.data.indexd.IndexedFile.index_document", record
     )
     mocker.add_mock(indexd_patcher)
 
 
 @pytest.fixture(scope="function")
@@ -552,12 +550,11 @@
             "created_date": "",
             "updated_date": "",
         }
 
     indexd_patcher = patch(
-        "fence.blueprints.data.indexd.IndexedFile.index_document",
-        record,
+        "fence.blueprints.data.indexd.IndexedFile.index_document", record
     )
     mocker.add_mock(indexd_patcher)
 
 
 @pytest.fixture(scope="session")
@@ -576,12 +573,11 @@
         "uploader": uploader_username,
         "created_date": "",
         "updated_date": "",
     }
     indexd_patcher = patch(
-        "fence.blueprints.data.indexd.BlankIndex.index_document",
-        record,
+        "fence.blueprints.data.indexd.BlankIndex.index_document", record
     )
     mocker.add_mock(indexd_patcher)
     return type("UploadIndexClient", (object,), {"record": record})
 
 
@@ -660,12 +656,11 @@
             "created_date": "",
             "updated_date": "",
         }
 
     indexd_patcher = patch(
-        "fence.blueprints.data.indexd.IndexedFile.index_document",
-        record,
+        "fence.blueprints.data.indexd.IndexedFile.index_document", record
     )
     mocker.add_mock(indexd_patcher)
 
 
 @pytest.fixture(scope="function")

--- tests/data/test_data.py
+++ blackened
@@ -339,16 +339,11 @@
     with pytest.raises(ValueError):
         response.json
 
 
 def test_blank_index_upload(
-    app,
-    client,
-    auth_client,
-    upload_indexd_client,
-    encoded_creds_jwt,
-    user_client,
+    app, client, auth_client, upload_indexd_client, encoded_creds_jwt, user_client
 ):
     headers = {
         "Authorization": "Bearer " + encoded_creds_jwt.jwt,
         "Content-Type": "application/json",
     }
```
</details></br>

This formatting comment was generated automatically by a script in [uc-cdis/wool](https://github.com/uc-cdis/wool).

File: fence/blueprints/data/indexd.py [hunk 1]
```diff
@@ -33,97 +40,110 @@
...
+        """
+        return self.index_document["did"]
+
+    @cached_property
+    def index_document(self):
+        index_url = self.indexd.rstrip("/") + "/index"
+        params = {"uploader": self.uploader}
+        indexd_response = requests.post(index_url, json=params)
+        if indexd_response.status_code != 200 and indexd_response.status_code != 201:
+            raise InternalError(
```
Comment (giangbui) [inline fence/blueprints/data/indexd.py]: You can make the error message more intuitive. The error message should be in `indexd_response.json()`

Comment (rudyardrichter) [reply fence/blueprints/data/indexd.py]: fixed in 9f583de

File: fence/blueprints/data/indexd.py [hunk 2]
```diff
@@ -33,97 +40,110 @@
...
-        # no protocol specified, return first location as signed url
-        elif len(self.indexed_file_locations) > 0:
-            signed_url = self.indexed_file_locations[0].get_signed_url(
-                action, expires_in, public_data=self.public
-            )
-        else:
-            # at this point, they haven't specified a protocol and we don't
-            # have any actual locations, error out
-            raise NotFound("Can't find any file locations.")
+class IndexedFile(object):
```
Comment (giangbui) [inline fence/blueprints/data/indexd.py]: Not inherit from IndexRecord?

File: setup.py [hunk 3]
```diff
@@ -10,7 +10,7 @@
         "boto>=2.36.0,<3.0.0",
         "botocore>=1.7,<1.9.0",
         "boto3>=1.5,<1.6",
-        "cirrus",
+        "cached_property>=1.5.1,<2.0.0" "cirrus",
```
Comment (giangbui) [inline setup.py]: need new line!

Comment (rudyardrichter) [reply setup.py]: good catch, fixed

File: fence/blueprints/data/blueprint.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    Return a presigned URL for use with uploading a data file.
+
+    See the documentation on the entire flow here for more info:
+
+        https://github.com/uc-cdis/cdis-wiki/tree/master/dev/gen3/data_upload
+
+    """
+    # make new record in indexd, with just the `uploader` field (and a GUID)
+    blank_index = BlankIndex()
```
Comment (philloooo) [inline fence/blueprints/data/blueprint.py]: can we create blank_index after request validation, so we don't create it if user doesn't provide filename?

File: fence/blueprints/data/indexd.py [hunk 5]
```diff
@@ -33,97 +40,106 @@
...
+                "received error from indexd trying to create blank record: {}".format(
+                    indexd_response.json()
+                )
+            )
+        return indexd_response.json()
 
-    def get_signed_url(self, protocol, action, expires_in):
-        if not self.public and not self.check_authorization(action):
-            raise Unauthorized("You don't have access permission on this file")
+    def make_signed_url(self, filename, expires_in=None):
```
Comment (philloooo) [inline fence/blueprints/data/indexd.py]: it takes a expires_in but that's never passed from the request handler?

File: fence/config-default.yaml [hunk 6]
```diff
@@ -303,6 +303,9 @@ S3_BUCKETS:
     cred: 'CRED1'
     role-arn: 'arn:aws:iam::role1'
 
+# this must be one of the first keys under `S3_BUCKETS`
```
Comment (Avantol13) [inline fence/config-default.yaml]: maybe add a little more detail about what this is used for? `a bucket where all user uploads go when interacting with the /data/upload endpoint` or something like that

File: openapis/swagger.yaml [hunk 7]
```diff
@@ -298,6 +298,44 @@ paths:
...
+        field in the new record.
+      security:
+        - OAuth2:
+          - user
+          - data
+      operationId: uploadBlank
+      parameters:
+        - name: filename
+          required: true
+          TODO
```
Comment (Avantol13) [inline openapis/swagger.yaml]: does it take other params now? Is this a TODO for the endpoint or for the docs :P

File: fence/config-default.yaml [hunk 8]
```diff
@@ -322,6 +325,11 @@ HTTP_PROXY:
 # example value: 'https://example.com/index'
 INDEXD: null
 
+# this is the username which fence uses to make authenticated requests to indexd
+INDEXD_USERNAME: 'gdcapi'
```
Comment (Avantol13) [inline fence/config-default.yaml]: I know this is super unintuitive and definitely not DRY, but there is a `test-fence-config.yaml` in the tests folder. It's okay if it's not the same as this one (since it'll just default the config to these values) but it does print some warnings about missing vars to the command line. 

tldr; might be better to add these new config vars into the test-fence-config.yaml as well.
